eduhelx_jupyterlab_student/handlers.py

```diff
@@ -281,48 +186,188 @@ async def post(self):
+    repo_root = context._compute_repo_root(course["name"])
+    try:
+        # We're just confirming that the repo root is a git repository.
+        # If it is, don't need to clone
+        get_git_repo_root(path=repo_root)
+    except InvalidGitRepositoryException:
+        # We're not going to bother checking if fork_cloned is False actually.
+        # If the repo isn't properly setup for any reason, we'll just rename
+        # whatever is using that directory name, then try to set it up again.
+        # If we fail at any point, just abort so the extension crashes.
+        
+        master_repository_url = course["master_remote_url"]
+        student_repository_url = student["fork_remote_url"]
+
+        # We absolutely don't want to allow overriding any existing files.
+        # To be extra careful, we will move the existing directory if it isn't empty.
+        if repo_root.exists() and list(repo_root.iterdir()) != [repo_root / ".git"]:
+            uniq_rname = str(repo_root) + "~{}"
+            while os.path.exists(uniq_rname.format(c)):
+                c += 1
+            repo_root.rename(uniq_rname.format(c))
+            # Recreate the directory
+            repo_root.mkdir()
+            await set_git_authentication(context, course, student)
+
+        """ Now we're working with a new, empty directory. """
+        try:
+            # This block should never fail. If it does, just abort.
+            init_repository(repo_root)
+            await set_git_authentication(context, course, student)
+            add_remote(UPSTREAM_REMOTE_NAME, master_repository_url, path=repo_root)
+            add_remote(ORIGIN_REMOTE_NAME, student_repository_url, path=repo_root)
+
+            @backoff.on_exception(backoff.constant, Exception, interval=2.5, max_time=15)
+            def try_fetch(remote):
+                fetch_repository(remote, path=repo_root)
+
+            # If either of these reach backoff and fail, just abort
+            try_fetch(ORIGIN_REMOTE_NAME)
+            checkout(f"{ MAIN_BRANCH_NAME }", path=repo_root)
+            try_fetch(UPSTREAM_REMOTE_NAME)
+
+            @backoff.on_exception(backoff.constant, Exception, interval=2.5, max_time=15)
+            async def mark_as_cloned():
+                await context.api.mark_my_fork_as_cloned()
+
+            # Mark the fork as cloned. If this fails, just abort.
+            await mark_as_cloned()
+        except Exception as e:
+            # At this point, we're removing a folder we just made, so no chance of deleting user data.
+            shutil.rmtree(repo_root)
+            raise e
+
+async def set_git_authentication(context: AppContext, course, student) -> None:
+    repo_root = context._compute_repo_root(course["name"])
+    student_repository_url = student["fork_remote_url"]
+
+    try:
+        get_git_repo_root(path=repo_root)
+        execute(["git", "config", "--local", "--unset-all", "credential.helper"], cwd=repo_root)
```

Comment:
How do you know you're currently in the `repo_root` as the cwd? If you answer with "because this is called by `clone_repo_if_not_exists` and we know that in that function you will be in the `repo_root` as the cwd", I don't think that's good, and we should make our functions pure. So I'd put another `execute` here changing the cwd to the repo_root first, if that's one of the preconditions of this command.

Reply:
We don't know if our actual cwd equals `repo_root` at this point (we don't care about our actual cwd), we're just executing the command from the cwd of `repo_root`. It would make for a huge mess if we added changing directory. The `cwd` parameter in execute doesn't have the side effect of actually changing the directory of the running program; for more details, it's a kwarg of `subprocess.Popen`.

Reply:
Gotcha. I guess that makes sense 😂.